### PR TITLE
[roku-high-1]: Defaults for installment loans

### PR DESCRIPTION
**Pull request by PowVT**
Per Roku comments: "In the function` _verifyDefaultedLoan`, if it's a loan with installments, it only checks if the missed installments is < 40% of total number of installments. This leads to scenarios when the borrower still keeps paying the 'minimum' interest, it will make the loan "non defaultable" and the lender would have no way to get his principal back even after the duration."

Fix: Logic was added to the `claim` function in LoanCore. This function now will return the loanId as Defaulted if the function call is made after the `dueDate`. The `dueDate` is `startDate + durationSecs`. This behavior mimics the default behavior for legacy loan types. 

Tests: 3 tests were added to the end of Installments.ts for better introspection of short term loans with minimal installments. The purpose of these tests are to prove out the logic described above.

### CODE REVIEW

File: test/Installments.ts [hunk 1]
```diff
@@ -2238,5 +2238,143 @@ describe("Installments", () => {
             expect(borrowerBalanceAfter.sub(borrowerBalanceBefore)).to.equal(0);
             expect(lenderBalanceAfter.sub(lenderBalanceBefore)).to.equal(1);
         });
+        it("Scenario: numInstallments: 4, durationSecs: 1 month. Borrower repays min 4 times and leaves. Lender tries to claim various times in duration.", async () => {
```
Comment (kkennis) [inline test/Installments.ts:2241]: Can you add a line of whitespace between all these tests?

File: contracts/LoanCore.sol [hunk 2]
```diff
@@ -243,14 +243,20 @@ contract LoanCore is
...
-        if (data.terms.numInstallments == 0) {
-            uint256 dueDate = data.startDate + data.terms.durationSecs;
+        // All loans types are able to be claimed after the loanId dueDate.
+        // First check if the loan has installments or the call is being made after the dueDate.
+        // In either of these cases, re-check the call is not before the dueDate
+        // since this was not checked prior for non-installment loan types. Only for installment type loans,
+        // a loan can be claimed prior to the dueDate. This can occur when greater than 40% on the total
+        // installment periods have been missed.
+        uint256 dueDate = data.startDate + data.terms.durationSecs;
+        if (data.terms.numInstallments == 0 || dueDate < block.timestamp) {
```
Comment (kkennis) [inline contracts/LoanCore.sol:253]: This looks a bit wonky....we are basically checking the opposite condition of what could be true `dueDate < block.timestamp`. I think this conditional logic can still be cleaned up.

I think we just need one check, that looks like:

```
if (block.timestamp > dueDate || data.terms.numInstallments > 0) {
    if (data.terms.numInstallments > 0) _verifyDefaultedLoan(...);
    
    // do claim...
} else {
   revert LC_NotExpired(...)
}
```

Comment (kkennis) [reply contracts/LoanCore.sol:253]: Goal would be to make the logic read as close to how you'd describe it in plain English. I'd probably say something like, "If the due date is passed or it's an installment loan, we can claim - but first, if it's an installment loan, verify the default. Otherwise, it's not expired".

Which matches with the code above

Comment (PowVT) [reply contracts/LoanCore.sol:253]: Thanks I really appreciate these comments. I agree the comments and code could be much more readable. I significantly shortened the comment blurb above the code, trying to better explain the conditions.

After, I made the change to the `if` statement and ran into an issue. Here is the scenario: an installment loan that is expired and the borrower has been making all their "minimum" (interest only) payments. For this scenario, the claim should happen automatically because the call is made after the `dueDate` and the loan is not paid off, but instead it enters the `if` statement and tries to check if the loan has missed more than 40% of the total installments. Where instead, that check doesn't need to happen due to automatic default. 

I think opting for explicitness in the `if` statement for `data.terms.numInstallments == 0` is important. It is a double check the loan terms are not tampered with in any way (which we know is not possible, but for any unknowns it makes our protocol predictable). Also, I swapped around the order of `block.timestamp > dueDate` to make that more readable the way you have it. Thanks again, let me know what you think.

**Review by kkennis — APPROVED** [review]
Got it, makes sense! Thanks for talking through the approach.
